### PR TITLE
Metronome accent patterns

### PR DESCRIPTION
**Pull request by pljones**
I'm raising this as a PR.

I still can't get the VST working.

When I load it into Reaper, it starts up okay and then shows this popup:
jamtaba-launch-popup
After clicking OK, it just vanishes, still:
jamtaba-launch-clicked-ok

Here's the debug output:
[JamTaba2.1.0-met.txt](https://github.com/elieserdejesus/JamTaba/files/1315384/JamTaba2.1.0-met.txt)

### CODE REVIEW

**Comment by elieserdejesus** [comment]
@pljones , try run **qmake** in QtCreator (click with right button in project name) to force a recompilation.   
 I'm suspecting your Configurator.h (the VERSION) is not updated.

**Comment by pljones** [comment]
Still getting the version error.  I have:
- deleted my local git checkout and cloned again
- deleted the output folder
- checked Configurator.h (says 2.1.0)
- run a Clean all
- run a qmake
- run a Rebuild all

**Comment by jonjamcam** [comment]
@pljones did you uninstall other jamtaba vst from your system. It could be recognizing an old version.

**Comment by pljones** [comment]
OK, uninstalling my day-to-day copy worked (I'd not upgraded for a while).

Previously this hasn't been an issue - I've been able to have an installed copy and my development copy has taken priority because it's later in the search path.

Still got the issue that I can't type space into the input field and I'm not getting change events from it, though.

**Comment by elieserdejesus** [comment]
> Basically, I'm having no luck at all with the UI, any help there would be appreciated.

I will check soon. The spacebar issue is related with VST, I need check the details with more time.

File: src/Common/gui/NinjamPanel.cpp [hunk 1]
```diff
@@ -256,9 +258,34 @@ int NinjamPanel::getPanSliderMaximumValue() const
...
+
+QString NinjamPanel::getAccentBeatsText() const
+{
+    return ui->lineEditAccentBeats->text();
+}
+
+void NinjamPanel::setAccentBeatsText(QString value)
+{
+    qCDebug(jtNinjamGUI) << "NinjamPanel::setCurrentAccentBeats " << value;
+    ui->lineEditAccentBeats->blockSignals(false);
```
Comment (elieserdejesus) [inline src/Common/gui/NinjamPanel.cpp]: @pljones , this is this problem. The block/unblock signals commands are inverted. Block first, and unblock in the end. Tested here and works fine after fix the block/unblock.

Comment (pljones) [reply src/Common/gui/NinjamPanel.cpp]: You're right - but that didn't fix the space bar issue (still goes to the host).

**Comment by elieserdejesus** [comment]
@pljones , lineEdits are *hacked* in VST plugin. The steps to get the accents line edit working are:

***
1 - Add a new parameter in NinjamPanel constructor:
image

***

2 - Change the NinjamRoomWindow accordinly:

image

***

3 - Add a 3rd parameter in **NinjamPanel::initializeCombos** and modify (install the modifier) in your lineEdit.

image

**Comment by pljones** [comment]
Ah... I was wondering what the bpm/bpi code was doing.  I'll give that a go, thanks!

**Comment by pljones** [comment]
Right, that's better.  However...
- the field is enabled even when "Custom..." isn't selected - that's not intended
- I clearly need to add a signal handler to catch the enter-completion

Still, it's much easier to test it now! 👍

**Comment by elieserdejesus** [comment]
@pljones 
> the field is enabled even when "Custom..." isn't selected - that's not intended

Please replace the lineEdit **setEnabled** with **setVisible**, the lineEdit will appear only to the **custom** option. The UI will be more clean.

Another UI improvement is use the  QLineEdit **placeHolderText** to explain the accents pattern.

**Comment by pljones** [comment]
OK, I'll to the visible as well - it will mean a gap in the UI when it's off, though.

**Comment by pljones** [comment]
Hm, I've broken something.  I had it so enter was updating and now you need to switch off and on again.

**Comment by elieserdejesus** [comment]
> The placeholder text doesn't appear to have any effect (nor does the tooltip - not that the tooltips work on other fields, either).

I suspect this is a VST only behavior, possible the textEdit hack is not considering the placehoder text and the tooltip.

**Comment by pljones** [comment]
I'll see what I can do (once I get a build environment for the standalone...).

**Comment by pljones** [comment]
Hi,

The latest on this is I'm stuck - as usual - with 64 bit portaudio not working with my ASIO drivers: it crashes on start up with no diagnostics.  (NOT a Jamtaba-specific issue as far as I'm aware.)

I didn't manage to get a build of portaudio I could use with Jamtaba, though.  If you could let me know how you build it, that could help.

Alternatively, just patch what's here now and I'll pull that back to my fork.

Thanks,

-- Peter

**Comment by elieserdejesus** [comment]
> I didn't manage to get a build of portaudio I could use with Jamtaba, though. If you could let me know how you build it, that could help.

@pljones , I compiled portaudio long time ago. But I think I changed just 2 things in the Visual Studio project (it's inside **build** folder). 

1 - I setted the **PA_USE_ASIO** preprocessor macro in project properties.
2 - Changed the link options to compile as **MD** (if I remember correctly the original project is compiling as MT). All JamTAba libs are compiled with **MD** flag and I'm packaging the **vc redistributable**.

***

> Alternatively, just patch what's here now and I'll pull that back to my fork.

Ok, I will integrate your code, thanks for your contribution again.

**Pull request closed without merging**